### ISSUE TITLE
'OpenCL Device' do not refresh on change of 'OpenCL Platform' (WebGUI / SETTINGS)

### ISSUE THREAD

**Issue by voidxno** [madMAx43v3r/mmx-node#268]
Got following OpenCL hardware:
- OpenCL Platform:`Intel(R) OpenCL Graphics`
  - OpenCL Device: `Intel(R) Graphics [0x7d67]`
- OpenCL Platform:`NVIDIA CUDA`
  - OpenCL Device: `NVIDIA GeForce RTX 3050`

**Issue:** In WebGUI / SETTINGS. When **switching between the two 'OpenCL Platform'**. The **'OpenCL Device' list is not refreshed**. It stays with device list it was initialized with (startup). Is cosmetic (to a certain degree). After mmx-node restart (needed anyway), Device is label'ed correctly.

Not a very important issue. Mostly cosmetic. Can be a case if 1x device on one platform, 2x devices on next. Might not be able to WebGUI select 2nd device on second platform (until restart of mmx-node).

Took a look at source:
- `Node.cpp` populates list of platforms in `opencl.platform_list`
- `Node.cpp` populates list of devices in `opencl.device_list` (but **only for currently active platform**)
- `settings.js` gets data through `WebAPI.cpp` and `/wapi/config/get`
- `settings.js` populates `opencl_platform_list` and `opencl_device_list`
- `settings.js` **no refresh of** `opencl_device_list`, **when drop-down** `opencl_platform_list`

Maybe solution:
- Let `Node.cpp` populate data with two-dim array with devices from all platforms
- Change `settings.js` to act on this available data

**Not a priority**. Just wanted an issue to refer to if comes up.

**Comment by madMAx43v3r** [comment]
Yeah this is currently a limitation...

**Comment by madMAx43v3r** [comment]
I think we should merge platform + device into a single config. Just show device name, and internally switch platform if needed.

**Comment by voidxno** [comment]
Yeah. Intel certainly on a run these days, just not in the right directions :-|

Wouldn't mind giving it a go, already looked at the code. Would something like this be a starting point ?

**Visual:**
- Remove WebGUI 'OpenCL Platform'
- Keep WebGUI 'OpenCL Device'
- Populate 'OpenCL Device' with merged entries
  - Sample (from above):
    - None (CPU)
    - Intel(R) Graphics [0x7d67] (Intel(R) OpenCL Graphics)
    - Intel(R) Arc(TM) A380 Graphics (Intel(R) OpenCL Graphics)
  - Sample (from above):
    - None (CPU)
    - NVIDIA GeForce RTX 3050 (NVIDIA CUDA)
    - NVIDIA GeForce GT 1030 (NVIDIA CUDA)
    - Intel(R) Graphics [0x7d67] (Intel(R) OpenCL Graphics)

**Logical:**
- Enumerate raw data found to one-dim arrays, representing total list of devices
- opencl.json / "platform": Keep friendly name
- Node.json / "opencl_device": Start using friendly name here too (not index)
- Match at startup config friendly names with list enumerated
- If match found: Using device and platform
- If no match: Fallback to first device in list (if any, else "None (CPU)")
- "None (CPU)" is represented with "-1", as it is now, (doubt any actual device has that name)

**Comment by madMAx43v3r** [comment]
I would keep the index for `opencl_device`. The only reason we use the string for platform is because the order is undefined, whereas the device order per platform is deterministic.

But if there are duplicate platform names, this wont work anyways... Need to find something else other than name or index to identify platform. This will require modifying `basic_opencl`.

Only list device name (simple/clean) -> Yes

**Comment by voidxno** [comment]
Duh. I just created a reverse problem.

What if two identical devices installed (like 2x RTX 3050). Will list them both. But would never be able to activate second. So need to config-store an 'index' in some form, yeah.

How to do it simple/elegant. Was hoping not triggering a change to `basic_opencl`.

**Comment by voidxno** [comment]
Just FYI: [clinfo.txt](https://github.com/user-attachments/files/17919102/clinfo.txt) attachment of both iGPU & Arc A380 present. Cannot find any identificator that differentiate between the two Intel platforms. Maybe OpenCL API has some, but not shown in clinfo.

And, cannot use index save-config for platform (per the undefined order).

**Comment by madMAx43v3r** [comment]
I think your idea of searching multiple platforms to find the selected device will work. So we need to store both device name + device index, and no platform name.

**Comment by madMAx43v3r** [comment]
> Was hoping not triggering a change to basic_opencl.

If it's just adding a new function, it's no problem. This is my library.

**Comment by madMAx43v3r** [comment]
Ok I would add `opencl_device_name` config. And then fallback to current method (plat name + dev index) if this config is not specified. In case `opencl_device_name` is specified then `opencl_device` becomes a relative index, instead of absolute, to specify which card to use in case of multiple duplicates.

The new `opencl.device_list_all` config needs to be an array of objects: `[{name: ..., index: <relative>, platform: <name>}, ...]`

And the WebGUI needs to be updated to display `opencl.device_list_all` correctly, and set both `opencl_device_name` and `opencl_device`, and maybe even `opencl.platform`. If `opencl.platform` is not specified we search through all, otherwise limit the search to the given platform name.

**Comment by voidxno** [comment]
Sounds sound:
- `platform` (opencl.json): No saved device name = legacy/migrate config value.
- `opencl_device` (Node.json): No saved device name = legacy/migrate config value. Else, -1: CPU, >=0: relative index.
- `opencl_device_name` (Node.json): New config value, device name to use (search for).

Manual config (no WebGUI):
- List all devices in .log at startup (enumeration), won't be many.

Goal:
- No problem with identical platform or device names.
- Legacy/migration values supported.
- Single clean new WebGUI selection list (all devices visible).

**Request**:
- Could you create a branch on mmx-node from current master ? Only change, define `opencl_device_name` (Node.json) config value (string) with your vnx magic.

**Comment by madMAx43v3r** [comment]
I pushed it on master.

**Comment by voidxno** [comment]
Even better, thx.

**Comment by voidxno** [comment]
<ins>Only noting/documenting for myself</ins>, if anyone else should observe it. No practical effect.

**Effect:** While testing [PR#270](https://github.com/madMAx43v3r/mmx-node/pull/270) my **Win10** installation **reports NVIDIA cards twice through OpenCL**.

Logically looks like this:

```
-- NVIDIA CUDA
   +-- NVIDIA GeForce RTX 3050
   +-- NVIDIA GeForce GT 1030
-- Intel(R) OpenCL Graphics
   +-- Intel(R) Graphics
-- NVIDIA CUDA
   +-- NVIDIA GeForce RTX 3050
   +-- NVIDIA GeForce GT 1030
```

Enum of OpenCL platforms give 2x separate NVIDIA CUDA platforms (different handles, but same properties), that gives that same hardware (different handles, but same hardware). Why that Win10 installation have gotten into that state (extra 'NVIDIA CUDA' OpenCL platform), unknown. Not observed same effect on my other Win10 platforms.

Also. AIDA64 and Geekbench on that Win10 platform, reports identical 'double' effect for OpenCL. Those 2x real NVIDIA hardware GPU's works fine, independent of selecting first or second instance.

For now, chalking it up to my Win10 installation getting into a weird OpenCL state. No need, and probably wrong getting [PR#270](https://github.com/madMAx43v3r/mmx-node/pull/270) to handle something that probably is state of OS issues.

**Comment by voidxno** [comment]
[PR#270](https://github.com/madMAx43v3r/mmx-node/pull/270) now merged, solves issues described.
